### PR TITLE
Added border curve style prop ("Squircle" effect - iOS only)

### PR DESCRIPTION
**Pull request by eric-edouard**
## Summary

NOTE: PR is based on https://github.com/facebook/react-native/pull/32017 which went stale for quite a long time but can now safely be closed 

Since iOS 13+, it is possible to change the corner curve property on iOS in order to smoothen border radius and make it more "rounded" (also called "squircle")
Here's an [article](https://medium.com/@arthurofbabylon/a-smooth-corner-radius-in-ios-54b80aa2d372) explaining in details what it is.
This property is also built in figma, but currently there is no way to implement this directly with react-native despite it being available natively on iOS.

Many open source react-native libraries were created in order to simulate this behaviour:
[react-native-super-ellipse-mask](https://github.com/everdrone/react-native-super-ellipse-mask)
[react-native-squircle-view](https://github.com/everdrone/react-native-squircle-view)
[react-native-figma-squircle](https://github.com/tienphaw/react-native-figma-squircle)

But they rely on creating an SVG shape with the smoothed corners and masking the view behind. This makes it not very performant (flickering on mounting was a common side-effect)

This PR aims at implementing the property natively.

PR for the docs update: https://github.com/facebook/react-native-website/pull/2785

## Changelog

[iOS] [Added] - Added `borderCurve` style prop for smooth border radius (squircle effect)

## Test Plan

We used the RNTester app and added an example with `cornerCurve ` set to `'continuous'` (only on iOS).

As the difference is quite subtle, we also made some more tests to better illustrate the difference (these are not in the RN-tester app):

IMG_0810

We overlapped two views with `position: absolute`, the one in the background has a red background and has `cornerRadius` set to `false`, and the one in the foreground is set to `true`. We can clearly see where the borders differs on the corners.

### CODE REVIEW

**Review by janicduplessis — APPROVED** [review]
Nice! PR looks good to me.

Sorry this is taking forever to merge

cc @cortinico

**Comment by janicduplessis** [comment]
@eric-edouard Did you test if this work with the Fabric renderer? I think RN tester uses it by default now.

Can you also rebase, I think CI was broken recently and was fixed here https://github.com/facebook/react-native/commit/94341f03809225f9ac97d7be0cfd08d10a62269c.

**Comment by janicduplessis** [comment]
I think similar logic could be used here https://github.com/facebook/react-native/blob/main/React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm#L582.

**Comment by eric-edouard** [comment]
@janicduplessis you're correct, this doesn't work with Fabric yet. Will take a look into it! Thanks

File: React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm [hunk 1]
```diff
@@ -580,6 +580,16 @@ - (void)invalidateLayer
     layer.borderColor = borderColor;
     CGColorRelease(borderColor);
     layer.cornerRadius = (CGFloat)borderMetrics.borderRadii.topLeft;
+    if (@available(iOS 13.0, *)) {
```
Comment (eric-edouard) [inline React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm:583]: Not sure about adding the switch case directly in there, should I extract this in a function ?

**Comment by eric-edouard** [comment]
@janicduplessis I've adapted the logic to have it work with Fabric, works well on my end with the same testing strategy as in the PR description.
I mostly duplicated the logic found in both BorderStyle (It being a string prop and an enum in native) and BorderRadius (as it affects corners)
it did seem like a lot to add in many places for such a small feature 😄 but couldn't see how to make it simpler.

File: React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm (hunk 1 above)
Comment (janicduplessis) [reply React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm:583]: I guess you could extract it to a function like this one https://github.com/facebook/react-native/blob/de72e4d62e62eb5f98b4e39f9e4e126d7a9237d8/React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm#L519

Comment (eric-edouard) [reply React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm:593]: @janicduplessis done!

**Comment by eric-edouard** [comment]
@janicduplessis quick bump 🙂

**Review by cipolleschi — APPROVED** [review]

**Comment by eric-edouard** [comment]
@cipolleschi rebasing seem to have broken the feature, will fix soon

**Comment by eric-edouard** [comment]
@cipolleschi rebased + tested again on RNTester app after merge, all good

**Comment by eric-edouard** [comment]
@cipolleschi quick bump!

File: React/Views/RCTBorderCurve.h [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  * This source code is licensed under the MIT license found in the
+  * LICENSE file in the root directory of this source tree.
+  */
+
+#import <Foundation/Foundation.h>
+
+typedef NS_ENUM(NSInteger, RCTBorderCurve) {
+	RCTBorderCurveContinuous = 0,
+	RCTBorderCurveCircular,
+};
```
Comment (cipolleschi) [inline React/Views/RCTBorderCurve.h:13]: Hi, could you add a new line in this file? Our internal linter is complaining about this.

**Review by cipolleschi — COMMENTED** [review]
Sorry if this took a while to be imported. Thanks again for the contribution.
Can I ask if you can rebase it and fix the small comment I left?

Besides these, it looks very good.

File: React/Views/RCTBorderCurve.h (hunk 2 above)
Comment (eric-edouard) [reply React/Views/RCTBorderCurve.h]: Hi, yes of course, should be ok now

**Comment by eric-edouard** [comment]
@cipolleschi No worries! Fixed the new line and aligned with main. Let me know if there anything else

**Pull request closed without merging**